examples2d/balls2.rs

```diff
@@ -80,7 +82,7 @@ fn main() {
     /*
      * Set up the testbed.
      */
-    let mut testbed = Testbed::new(world);
+    let mut testbed = Testbed::new(Arc::new(Mutex::new(world)));
```

Comment:
This should be avoided for several reasons:

1.  It means all the testbed examples will have to be modified.
2.  This complicates the example with something that is useful only in some corner-use-cases.
2. I don't want users looking at the examples to think using an `Arc<Mutex<...>>>` is always necessary. (Someone quickly skimming through the examples just to see what they look like will notice the `use` of `Arc` and `Mutex` so they may end up concluding too quickly).

An alternative is to keep the `Testbed::new(world)` as it is, and have a method from the testbed to retrieve the shared world: `testbed.world()`.

Reply:
My alternative take on this was to make different instantiaters to pass world to Testbed, from my WIP local branch:

```rust
/// This trait is designed to allow choosing implementation of underlying storing of World: shared between threads or owned only by Testbed.
pub trait WorldOwner {
    fn get<'a: 'b, 'b>(&'a mut self) -> Box<DerefMut<Target = World<f32>> + 'b>;
}

pub struct WorldOwnerShared {
    world: Arc<Mutex<World<f32>>>,
}

impl WorldOwnerShared {
    pub fn new(w: World<f32>) -> WorldOwnerShared {
        WorldOwnerShared {world: Arc::new(Mutex::new(w))}
    }
}

impl WorldOwner for WorldOwnerShared {
    fn get<'a: 'b, 'b>(&'a mut self) -> Box<DerefMut<Target = World<f32>> + 'b> {
        Box::new(self.world.lock().unwrap())
    }
}
// TODO: simpler WorldOwnerExclusive implementation not requiring mutex
```

that would impact my use case example as 
```rust
let mut testbed = Testbed::new(Testbed::WorldOwnerShared(world)));
```
And existing simpler examples as 
```rust
let mut testbed = Testbed::new(Testbed::WorldOwnerExclusive(world)));
```

*Maybe the world itself could implement WorldOwner, so the existing examples wouldn't be impacted at all ? 💃*

Your suggestion makes it difficult or overkill to share `TestBed` with `specs::System`, in order to retrieve `nphysics::World`. **Please tell me** if my lead is incorrect so I don't spend time in a incorrect direction.

All this discussion is underlying to the goal of objective 
>allow to avoid heavy Arc<Mutex<nphysics::World>> if not required by (library) user."

from the **TODO List**

Reply:
> Your suggestion makes it difficult or overkill to share TestBed with specs::System, in order to retrieve nphysics::World.

There is no need to share the testbed with specs. What I was thinking of would mean your [example](https://github.com/Vrixyz/testbed-callback/blob/1-arc-mutex/src/main.rs#L122-L136) would look like that:

```rust
    // Initialize the testbed first.
    let mut testbed = Testbed::new(physics_world);

    let ecs_world = std::cell::RefCell::new(specs::World::new());

    // In the following, we have testbed.world() -> Arc<Mutex<World<f32>>
    ecs_world
        .borrow_mut()
        .add_resource(testbed.world());
    let dispatcher = DispatcherBuilder::new()
        .with(DummySystem::new(), "dummy_system", &[])
        .build();

    let dispatcher = std::cell::RefCell::new(dispatcher);

);
```
Basically, the testbed is responsible for creating the `Arc` internally.

-----

If think your solution based on `WorldOwner` is better than what I proposed above. The testbed could be parameterized by: `Testbed<W: WorldOwner>`, and instead of returning `Box<DerefMut<Target = World<f32>> + 'b>>`, the `.get` method you propose could return `impl DerefMut<Target = World<f32>> + 'b>` to avoid boxing.

As you proposed, both `World`, `Arc<RwLock<World>>` and `Arc<Mutex<World>>` should implement this `WorldOwner` trait so no modification of any testbed example is required (even when user callbacks are used since they would become `Fn(&mut T, &mut GraphicsManager, f32)`.

I don't think you will be able to make a solution based on `From<nphysics::World>` without requiring the user to add explicit type annotation for the testbed though.

Reply:
What I thought about bypasses modifications of existing examples :tada: : 

```rust
pub fn set_world(&mut self, world: World<f32>) {
    self.set_world_owner(Box::new(WorldOwnerExclusive::from(world)));
}

/// Lower level setter allowing for shared World behaviour
pub fn set_world_owner(&mut self, world: Box<WorldOwner>) ...
```

I tried experiencing with `Into` and `From` for World and `Arc<RwLock>`, but didn't succeed in making the situation better, I'll stop there my experiments on `set_world` API modifications and consider it OK enough.

Reply:
I did try to make `get` return a `impl WorldOwned`, but I failed at making it compile again, so I'll let this as an exercise to whoever is more confortable with that.